---
title: "Diff_Exp_Peaks.Rmd"
output: html_document
date: "2024-05-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Differential Expression Analysis of Peaks using Limma Voom
## In order to get the top table we ran the R script on an HPC due to large number of fragments.

## Load libraries 
```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
library(EnhancedVolcano)
library(gridExtra)
library(ComplexHeatmap)
```

```{r}
## Load data 
atac <- readRDS(here::here("Data", "ATAC_signac.rds"))

## Extract counts 
eset <- atac@assays$ATAC@counts
ids <- unlist(sapply(strsplit(colnames(eset), "_"), function(i) i[[1]]))
sex <- factor(atac@meta.data$sex, levels = c("Male", "Female"))
drug <- factor(atac@meta.data$drug, levels = c("Saline", "Fentanyl"))
cell_labels <- atac@meta.data$celltype

# Remove the specific cell type
excluded_cell_type <- "Ex_Onecut3_Spag16_3"
cell_labels <- cell_labels[cell_labels != excluded_cell_type]
eset <- eset[, cell_labels != excluded_cell_type]

eset <- as.matrix(eset)
library(parallel)
cl <- makeCluster(detectCores())
clusterExport(cl, c("cell_labels", "sex", "drug", "eset"))
top <- parLapply(cl, names(table(cell_labels)), function(cell){
  library(limma);
  cat(cell, fill = TRUE)
  keep <- cell_labels %in% cell
  counts <- eset[, keep]
  dat <- data.frame(sex = sex[keep],
                    drug = drug[keep],
                    subj = gsub("_.*", "", colnames(counts)))
  
  design <- model.matrix(~sex*drug, data = dat)
  vobj_tmp <- voom(counts, design, plot = FALSE)
  dupcor <- duplicateCorrelation(vobj_tmp, design, block = dat$subj)
  vobj <- voom(counts, design, plot = FALSE, block = dat$subj, correlation = dupcor$consensus)
  dupcor <- duplicateCorrelation(vobj, design, block = dat$subj)
  fitDupCor <- lmFit(vobj, design, block = dat$subj, correlation = dupcor$consensus)
  fitDupCor <- eBayes(fitDupCor)
  top_FvsM <- topTable(fitDupCor, coef = "sexFemale", adjust.method = "BH", number = nrow(fitDupCor))
  top_FvsM$comparison <- "Female_vs_Male"
  top_FvsM$gene <- rownames(top_FvsM)
  top_SvsF <- topTable(fitDupCor, coef = "drugFentanyl", adjust.method = "BH", number = nrow(fitDupCor))
  top_SvsF$comparison <- "Saline_vs_Fentanyl"
  top_SvsF$gene <- rownames(top_SvsF)
  top_int <- topTable(fitDupCor, coef = "sexFemale:drugFentanyl", adjust.method = "BH", number = nrow(fitDupCor))
  top_int$comparison <- "drug x sex interaction"
  top_int$gene <- rownames(top_int)
  top <- rbind(top_FvsM, top_SvsF, top_int)
  top$cell <- cell
  top }) 
stopCluster(cl)

top <- Reduce("rbind", top)

## Save top table as rds 
saveRDS(top, file = "top_atac.rds")
```


## load all 3 top tables (atac, gene activity and gene expression)
```{r}
# gene activity top table 
top_ga <- readRDS(here::here("Data", "top_ga.rds"))

# top table for just fragments 
top_atac <- readRDS(here::here("Data", "top_atac.rds"))

# top table using RNA counts 
top_rna <- readRDS(here::here("Data", "top_gex.rds"))
```

## Find closest genes to peaks
```{r}
# step 1: run closest feature that gives you the closest genes using the regions from atac object 
closest_genes <- ClosestFeature(atac, regions = granges(atac))

# Step 2: Filter differentially expressed peaks top table
differentially_expressed_peaks <- top_atac[top_atac$comparison == "drug x sex interaction" & top_atac$adj.P.Val < 0.01, ]

# Rename the 'gene' column to 'regions' as they are actully regions
differentially_expressed_peaks <- differentially_expressed_peaks %>%
  rename(regions = gene)

# Step 3: Left join query regions (from closest_genes) with regions that are differentially expressed (in differentially expressed peaks) and add the genes from the gene col into top table. 
differentially_expressed_peaks <- left_join(differentially_expressed_peaks, 
                         closest_genes %>% select(query_region, gene_name), 
                         by = c("regions" = "query_region"))

# Step 4: Filter differentially expressed genes top (top_gex) table by comparison and adjusted p value (< 0.01)
differentially_expressed_genes <- top_rna[top_rna$comparison == "drug x sex interaction" & top_rna$adj.P.Val < 0.01, ]
```


```{r}
# Step 5: Create heatmap with the x axis being the different cell types and the columns being number of genes that are only atac, number of genes that intersect and number of genes from gex. library(dplyr)


# Define the categories
categories <- c("Only ATAC", "Intersect", "Only GEX")

# Group by cell type and summarize
counts_df <- differentially_expressed_peaks %>%
  group_by(cell) %>%
  summarise(
    only_atac_genes = sum(!gene_name %in% differentially_expressed_genes$gene),
    intersect_genes = sum(gene_name %in% differentially_expressed_genes$gene),
    only_gex_genes = sum(!differentially_expressed_genes$gene %in% gene_name)
  ) %>%
  ungroup()

# Convert the data frame to a matrix for the heatmap
counts_matrix <- as.matrix(counts_df[, -1])
rownames(counts_matrix) <- counts_df$cell

# Create the heatmap
Heatmap(counts_matrix,
        name = "Number of Genes",
        row_title = "Cell Types",
        column_title = "Gene Categories",
        show_row_names = TRUE,
        show_column_names = TRUE,
        column_names_side = "bottom")

```
