---
title: "EnrichR.Rmd"
output: html_document
date: "2024-05-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# EnrichR: Find enriched pathways using differentially expressed genes 

## Load top table 
```{r}
#Load in the top table from Differential expression analysis 
top <- readRDS("/Users/rishikadaswani/Desktop/Fentanyl_Multiomics/Data/top.rds")
```

## Filtering top table 
```{r}
# filter by comparison 
top_dxs <- top[top$comparison == "drug x sex interaction", ]

# filter by adj p.value 
top_dxs <- top_dxs[top_dxs$adj.P.Val < 0.01, ]

# filter by cell type 
Chol_Isl2_Gata6_4 <- top_dxs %>% filter(cell == "Chol_Isl2_Gata6_4") %>% filter(!grepl("^Mt-", gene))
Chol_Tbx20 <- top_dxs %>% filter(cell == "Chol_Tbx20_S100a11_8") %>% filter(!grepl("^Mt-", gene))
Ex_C1ql4 <- top_dxs %>% filter(cell == "Ex_C1ql4_Asb4") %>% filter(!grepl("^Mt-", gene))
Ex_Dmrta2 <- top_dxs %>% filter(cell == "Ex_Dmrta2_Prkcq_4") %>% filter(!grepl("^Mt-", gene))
Ex_Fezf2 <- top_dxs %>% filter(cell == "Ex_Fezf2_Fbln1") %>% filter(!grepl("^Mt-", gene))
Ex_Fefz2_Fos <- top_dxs %>% filter(cell == "Ex_Fezf2_Fibcd1_Fos_Tnn") %>% filter(!grepl("^Mt-", gene))
Ex_Lhx9 <- top_dxs %>% filter(cell == "Ex_Lhx9_Tyrp1_4") %>% filter(!grepl("^Mt-", gene))
Fibro_Eya2 <- top_dxs %>% filter(cell == "Fibro_Eya2_Slc47a1_2") %>% filter(!grepl("^Mt-", gene))
Inh_Nkx2 <- top_dxs %>% filter(cell == "Inh_Nkx2-2_Glp1r_Trh_1") %>% filter(!grepl("^Mt-", gene))
Inh_Six3 <- top_dxs %>% filter(cell == "Inh_Six3_Cela1_1") %>% filter(!grepl("^Mt-", gene))
```

## Load enrichR library 
```{r}
library(enrichR)
websiteLive <- getOption("enrichR.live")
if (websiteLive) {
    listEnrichrSites()
    setEnrichrSite("Enrichr") 
}

if (websiteLive) dbs <- listEnrichrDbs()

```

## Running EnrichR Pathway Analysis 

```{r}
#View and select databases 
dbs <- c("Allen_Brain_Atlas_10x_scRNA_2021", "KEGG_2019_Mouse", "RNAseq_Automatic_GEO_Signatures_Mouse_Down", "RNAseq_Automatic_GEO_Signatures_Mouse_Up")

#Run enrichR for each cell type 

enriched_Chol_Isl2_Gata6_4 <- enrichr(Chol_Isl2_Gata6_4$gene, dbs)
enriched_Chol_Tbx20 <- enrichr(Chol_Tbx20$gene, dbs)
enriched_Ex_C1ql4 <- enrichr(Ex_C1ql4$gene, dbs)
enriched_Ex_Dmrta2 <- enrichr(Ex_Dmrta2$gene, dbs)
enriched_Ex_Fefz2_Fos <- enrichr(Ex_Fefz2_Fos$gene, dbs)
enriched_Fibro_Eya2 <- enrichr(Fibro_Eya2$gene, dbs)
enriched_Inh_Nkx2 <- enrichr(Inh_Nkx2$gene, dbs)

```

## Plot enrichR output 
```{r}
if (websiteLive) {
    plotEnrich(enriched_Ex_C1ql4[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value")
}
```



