---
title: "EnrichR.Rmd"
output: html_document
date: "2024-05-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# EnrichR: Find enriched pathways using differentially expressed genes 

## Load top table 
```{r}
#Load in the top table from Differential expression analysis 
top <- readRDS(here::here("Data", "top_gex.rds"))
```

## Filtering top table 
```{r}
# filter by adj p.value 
top <- top[top$p_adj.loc < 0.3, ]

# filter by cell type 
Ex_C1ql4 <- top[top$cluster_id == "Ex_C1ql4_Drd1_3", ]
Ex_Plekhg1 <- top[top$cluster_id == "Ex_Plekhg1_Nox4_Kcp_2", ]
Ex_Fibcd1 <- top[top$cluster_id == "Ex_Fibcd1_Psrc1_4", ]
Ex_Tfap2b <- top[top$cluster_id == "Ex_Tfap2b_Ccdc172_Calcr", ]
Fibro_Eya2 <- top[top$cluster_id == "Fibro_Eya2_Ccdc192_2", ]
Inh_Frm7 <- top[top$cluster_id == "Inh_Frm7_Otop1_1", ]
```

## Load enrichR library 
```{r}
library(enrichR)
websiteLive <- getOption("enrichR.live")
if (websiteLive) {
    listEnrichrSites()
    setEnrichrSite("Enrichr") 
}

if (websiteLive) dbs <- listEnrichrDbs()

```

## Running EnrichR Pathway Analysis 

```{r}
#View and select databases 
dbs <- c("Allen_Brain_Atlas_10x_scRNA_2021", "KEGG_2019_Mouse", "RNAseq_Automatic_GEO_Signatures_Mouse_Down", "RNAseq_Automatic_GEO_Signatures_Mouse_Up")

#Run enrichR for each cell type 

enriched_Ex_C1ql4 <- enrichr(Ex_C1ql4$gene, dbs)
enriched_Ex_Fibcd1 <- enrichr(Ex_Fibcd1$gene, dbs)
enriched_Ex_Plekhg1 <- enrichr(Ex_Plekhg1$gene, dbs)
enriched_Ex_Tfap2b <- enrichr(Ex_Tfap2b$gene, dbs)
enriched_Fibro_Eya2 <- enrichr(Fibro_Eya2$gene, dbs)
enriched_Inh_Frm7 <- enrichr(Inh_Frm7$gene, dbs)

```

## Plot enrichR output 
```{r}
if (websiteLive) {
    plotEnrich(enriched_Ex_Tfap2b[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value")
}
```



