---
title: "atac_signac.Rmd"
output: html_document
date: "2024-05-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Signac: Analyzing scATAC-seq 

## Load libraries 
```{r}
#load libraries
library(Seurat)
library(Signac)
library(EnsDb.Rnorvegicus.v75)
library(GenomicRanges)
```

## Load seurat object 
```{r}
atac <- readRDS(here::here ("Data", "GEX_ATAC_Seur.rds"))
DefaultAssay(atac) <- "ATAC"
```


# Create a gene activity matrix 
```{r}
gene.activities <- GeneActivity(atac)

# add the gene activity matrix to the Seurat object as a new assay and normalize it
atac[['GA']] <- CreateAssayObject(counts = gene.activities)
atac <- NormalizeData(
  object = atac,
  assay = 'GA',
  normalization.method = 'LogNormalize',
  scale.factor = median(atac$nCount_RNA)
)
```

```{r}
# sctransform GA counts 
DefaultAssay(atac) <- "GA"
atac <- SCTransform(atac)
```


## Call Peaks 
```{r}
# in order for call peaks to run, you must have MACS2 which can be downloaded by using the conda - "conda install bioconda::macs2" or pip - "pip install MACS2". 
# peaks <- CallPeaks(
#   object = atac
# )
```

## Coverage Plot 
```{r}
# CoveragePlot(
#   object = atac,
#   region = "1-573693-574003",
#   ranges = peaks
# )
```

## Save rds 
```{r}
saveRDS(atac, file = here::here("Data", "GEX_ATAC_GA.rds"))
```


