---
title: "atac_signac.Rmd"
output: html_document
date: "2024-05-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Signac: Analyzing scATAC-seq 

## Load libraries 
```{r}
#load libraries
library(Seurat)
library(Signac)
library(EnsDb.Rnorvegicus.v75)
library(GenomicRanges)
```

## Load ATAC seurat object from qc_metrics rmd 
```{r}
atac <- readRDS("/Users/rishikadaswani/Desktop/Fentanyl_Multiomics/Data/ATAC_subset.rds")
```

## Normalization and linear dimensional reduction 
```{r}
atac <- RunTFIDF(atac) # normalization 
atac <- FindTopFeatures(atac, min.cutoff = 5) # find top features 
atac <- RunSVD(atac) # dimensionality reduction 

DepthCor(atac)
```

## Non-linear dimensionality reduction technique 
```{r}
atac <- RunUMAP(object = atac, reduction = "lsi", dims = 2:40)
atac <- FindNeighbors(object = atac, reduction = "lsi", dims = 2:40)
atac <- FindClusters(object = atac, algorithm = 3)
```

## Visualize cluster 
```{r}
DimPlot(object = atac, label = TRUE)
```


# Create a gene activity matrix 
```{r}
gene.activities <- GeneActivity(atac)

# add the gene activity matrix to the Seurat object as a new assay and normalize it
atac[['RNA']] <- CreateAssayObject(counts = gene.activities)
atac <- NormalizeData(
  object = atac,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(atac$nCount_RNA)
)
```

## Call Peaks 
```{r}
# in order for call peaks to run, you must have MACS2 which can be downloaded by using the conda - "conda install bioconda::macs2" or pip - "pip install MACS2". 
peaks <- CallPeaks(
  object = atac
)
```

## Coverage Plot 
```{r}
CoveragePlot(
  object = atac,
  region = "1-573693-574003",
  ranges = peaks
)
```

# Add cell labels from MultVI to atac object 

## Load in seurat object 
```{r}
gex_atac <- readRDS("/Users/rishikadaswani/Desktop/Fentanyl_Multiomics/Data/GEX_ATAC_Seur.rds")
```

## Extract cell labels 
```{r}
cell_types <- as.character(gex_atac@active.ident)
```

## add cell types into atac dataset 
```{r}
atac <- AddMetaData(object = atac, metadata = cell_types, col.name = "celltype")
```

# Add meta data (drug and sex)
```{r}
demo <- data.frame(ids = c("r3", "r4", "r15", "r16"),
                       drug = c("Saline", "Fentanyl", "Fentanyl", "Saline"), 
                       sex = c("Male", "Male", "Female", "Female"))
demo_10x <- atac@meta.data %>% tibble::rownames_to_column("ids") %>% mutate(ids = gsub("_.*", "", ids))
demo_10x <- demo_10x %>% left_join(demo, by = "ids")
new_metadata <- demo_10x[, c("drug", "sex")]
atac <- AddMetaData(object = atac, metadata = new_metadata)
```

## Save rds 
```{r}
saveRDS(atac, file = here::here("Data", "ATAC_signac.rds"))
```


