---
title: "Diff_ExpMuscat_ATAC"
output: html_document
date: "2024-06-06"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Differential State Analysis with `muscat` (ATAC Counts)

## Load packages 
```{r}
library(dplyr)
library(ggplot2)
library(limma)
library(muscat)
library(purrr)
library(scater)
library(sctransform)
library(UpSetR)
library(sceasy)
library(SingleCellExperiment)
library(Seurat)
```

## Load data and create SCE object 
```{r}
# load seurat object which contains gene activity matrix in "GA" assay 
atac <- readRDS(here::here("Data", "GEX_ATAC_Seur.rds"))

# extract GA counts 
atac_counts <- GetAssayData(atac, assay = "ATAC", slot = "counts")

# extract metadata 
metadata = atac@meta.data

# create sce object 
sce <- SingleCellExperiment(
  assays = list(counts = atac_counts),
  colData = metadata,
  metadata = metadata
)

sce
```

## Pre-processing 
```{r}
# remove undetected genes
sce <- sce[rowSums(counts(sce) > 0) > 0, ]
dim(sce)

# calculate per-cell quality control (QC) metrics
qc <- perCellQCMetrics(sce)

# remove cells with few or many detected genes
ol <- isOutlier(metric = qc$detected, nmads = 2, log = TRUE)
sce <- sce[, !ol]
dim(sce)

# remove lowly expressed genes 
sce <- sce[rowSums(counts(sce) > 1) >= 10, ]
dim(sce)

# compute sum-factors & normalize
sce <- computeLibraryFactors(sce)
sce <- logNormCounts(sce)

# perform variance stabilization transformation using the sctransform package 
#assays(sce)$vstresiduals <- vst(counts(sce), verbosity = FALSE)$y
```

## Data preparation 
```{r}
(sce <- prepSCE(sce, 
    kid = "aba_classification", # subpopulation assignments
    gid = "condition",  # group IDs (ctrl/stim)
    sid = "SampleID",   # sample IDs (ctrl/stim.1234)
    drop = TRUE))  # drop all other colData columns

# store cluster + sample IDs as well as number of clusters and samples 
nk <- length(kids <- levels(sce$cluster_id))
ns <- length(sids <- levels(sce$sample_id))
names(kids) <- kids; names(sids) <- sids
```

## Data overview 
```{r}
# nb. of cells per cluster-sample
df <- t(table(sce$cluster_id, sce$sample_id))
df

```

```{r}

gplots::heatmap.2(df/rowSums(df), trace="none",
                  margins = c(15, 10))
```

## Differential state (DS) analysis 
```{r}
# aggregation of single-cell to pseudo-bulk data 
pb <- aggregateData(sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id"))

# one sheet per subpopulation
assayNames(pb)

# pseudobulks for 1st subpopulation
t(head(assay(pb)))

# pseudobulk-level MDS plot 
(pb_mds <- pbMDS(pb))

# use very distinctive shaping of groups & change cluster colors
pb_mds <- pb_mds + 
  scale_shape_manual(values = c(21, 4)) +
  scale_color_manual(values = c(RColorBrewer::brewer.pal(8, "Set1"), RColorBrewer::brewer.pal(7, "Set2"), RColorBrewer::brewer.pal(6, "Set3")))
# change point size & alpha
pb_mds$layers[[1]]$aes_params$size <- 5
pb_mds$layers[[1]]$aes_params$alpha <- 0.6

pb_mds <- pb_mds + 
  geom_line(aes(group = interaction(cluster_id, group_id)), color = "grey", linetype = "dashed")

pb_mds <- pb_mds + ggtitle("Pseudobulk-level multidimensional scaling (MDS) plot for ATAC")

pb_mds
```

## Sample level analysis
```{r}
# construct design & contrast matrix
ei <- metadata(sce)$experiment_info
ei$group_id <- relevel(ei$group_id, ref = "Saline")
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("Fentanyl-Saline", levels = mm)

# run DS analysis
res <- pbDS(pb, design = mm, contrast = contrast, method="limma-voom")

# acess results table
tbl <- res$table$`Fentanyl-Saline`
```

## Handling results 
```{r}
# results filtering and overview 
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil <- lapply(tbl, function(u) {
  u_filtered <- subset(u, p_adj.loc < 0.3)
  u_filtered <- dplyr::arrange(u_filtered, p_adj.loc)
  return(u_filtered)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil, nrow, numeric(1))
p_de <- format(n_de / nrow(sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# calculate expression frequencies 
frq <- calcExprFreqs(sce, assay = "counts", th = 0)
# one sheet per cluster
assayNames(frq)
gids <- levels(sce$group_id)
frq10 <- vapply(as.list(assays(frq)), 
  function(u) apply(u[, gids] > 0.1, 1, any), 
  logical(nrow(sce)))
t(head(frq10))

```

## No differentially expressed fragments found so analysis was stopped here. However, the following steps can be done if significant results were found. 

```{r}
# generation of top table 
topfil_ATAC <- do.call(rbind, tbl_fil)
top_ATAC <- do.call(rbind, tbl)

# plot top table 
topfil_ATAC %>%
  ggplot(aes(x = logFC, y = -log10(p_val), color = cluster_id, label = gene)) +
  geom_point() +
  geom_text_repel() +
  theme_minimal() +
  labs(
    title = "Volcano Plot (ATAC)",
    x = "Log Fold Change",
    y = "-Log10 P-value",
    color = "Cluster"
  )

```

## Find the closest genes to the differentially expressed peaks 
```{r}
DefaultAssay(atac) <- "ATAC"

# run closest feature that gives you the closest genes using the regions from atac object 
closest_genes <- ClosestFeature(atac, regions = granges(atac))

# rename the 'gene' column to 'regions' as they are actually regions
topfil_ATAC <- topfil_ATAC %>%
  dplyr::rename(regions = gene)

# left join query regions (from closest_genes) with regions that are differentially expressed (in differentially expressed peaks) and add the genes from the gene col into top table. 
topfil_ATAC <- left_join(topfil_ATAC, 
                         closest_genes %>% select(query_region, gene_name), 
                         by = c("regions" = "query_region"))
```


## Visualize results 
```{r}
# between cluster concordance 
de_gs_by_k <- map(tbl_fil, "gene")
upset(fromList(de_gs_by_k))

# sample level viz: Pseudobulk heatmaps

# top-5 DS genes per cluster
pbHeatmap(sce, res, top_n = 5)
pbHeatmap(example_sce, res_0, top_n = 5)
```

```{r}
group <- factor(rep(c("fent", "sal"), each = 2))
plot(voom(assays(pb)$Endo_Flt1_Lrg1_1)$E["Lrrtm4", ] ~ group)

plot(assays(pb)$Endo_Flt1_Lrg1_1["Lrrtm4", ] ~ group)
```

## Save top table 
```{r}
saveRDS(topfil_ATAC, file = here::here("Data", "topfil_ATAC.rds"))
saveRDS(top_ATAC, file = here::here("Data", "top_ATAC.rds"))
```


```{r}
sessionInfo()
```
