---
title: "Diff_ExpMuscat_RNA"
output: html_document
date: "2024-06-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Differential State Analysis with `muscat` (RNA counts)

## Load packages 
```{r}
library(dplyr)
library(ggplot2)
library(limma)
library(muscat)
library(purrr)
library(scater)
library(sctransform)
library(UpSetR)
library(SingleCellExperiment)
library(Seurat)
```

## Load data and create SCE object 
```{r}
# Load seurat object 
rna_atac <- readRDS(here::here("Data", "GEX_ATAC_Seur.rds"))

# extract SCT counts 
sct_counts <- GetAssayData(rna_atac, assay = "SCT", layer = "counts")

# extract metadata 
metadata <- rna_atac@meta.data

# create singleCellExperiment object 
sce <- SingleCellExperiment(
  assays = list(counts = sct_counts),
  colData = metadata,
  metadata = metadata
)

sce 

```

## Pre-processing 
```{r}
# remove undetected genes
sce <- sce[rowSums(counts(sce) > 0) > 0, ]
dim(sce)

# calculate per-cell quality control (QC) metrics
qc <- perCellQCMetrics(sce)

# remove cells with few or many detected genes
ol <- isOutlier(metric = qc$detected, nmads = 2, log = TRUE)
sce <- sce[, !ol]
dim(sce)

# remove lowly expressed genes 
sce <- sce[rowSums(counts(sce) > 1) >= 10, ]
dim(sce)

# compute sum-factors & normalize
sce <- computeLibraryFactors(sce)
sce <- logNormCounts(sce)

```

## Data preparation 
```{r}
(sce <- prepSCE(sce, 
    kid = "aba_classification", # subpopulation assignments
    gid = "drug",  # group IDs (ctrl/stim)
    sid = "orig.ident", # sample IDs (ctrl/stim.1234)
    drop = TRUE))  # drop all other colData columns

# store cluster + sample IDs as well as number of clusters and samples 
nk <- length(kids <- levels(sce$cluster_id))
ns <- length(sids <- levels(sce$sample_id))
names(kids) <- kids; names(sids) <- sids
```

## Data overview 
```{r}
# nb. of cells per cluster-sample
df <- t(table(sce$cluster_id, sce$sample_id))
df

```

## Differential state (DS) analysis 
```{r}

# aggregation of single-cell to pseudo-bulk data 
pb <- aggregateData(sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id"))

# one sheet per subpopulation
assayNames(pb)

# pseudobulks for 1st subpopulation
t(head(assay(pb)))

# pseudobulk-level MDS plot 
(pb_mds <- pbMDS(pb))

# use very distinctive shaping of groups & change cluster colors
pb_mds <- pb_mds + 
  scale_shape_manual(values = c(17, 4)) +
  scale_color_manual(values = c(RColorBrewer::brewer.pal(8, "Set1"), RColorBrewer::brewer.pal(7, "Set2"), RColorBrewer::brewer.pal(6, "Set3")))
# change point size & alpha
pb_mds$layers[[1]]$aes_params$size <- 5
pb_mds$layers[[1]]$aes_params$alpha <- 0.6

pb_mds <- pb_mds + 
  geom_line(aes(group = interaction(cluster_id, group_id)), color = "grey", linetype = "dashed") + ggtitle("Pseudobulk-level multidimensional scaling (MDS) plot for RNA")

pb_mds
```

## Sample level analysis
```{r}
# construct design & contrast matrix
ei <- metadata(sce)$experiment_info
ei$group_id <- relevel(ei$group_id, ref = "Saline")
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("Fentanyl-Saline", levels = mm)

# run DS analysis
res <- pbDS(pb, design = mm, contrast = contrast, method="limma-voom", min_cells = 3)

# acess results table
tbl <- res$table$`Fentanyl-Saline`
```

## plot results

```{r}
df <- lapply(tbl, function(i){
  i %>% 
    arrange(p_val) %>% 
    mutate(n = 1:n())
}) %>% 
  do.call(rbind, .)

df %>% 
  ggplot(aes(x = n, y = p_adj.loc, color = cluster_id)) +
  geom_line() +
  scale_x_log10() +
  theme_classic() +
  facet_wrap(~cluster_id)
```


## Handling results 
```{r}
# results filtering and overview 
# filter FDR < 30% & sort by adj. p-value
tbl_fil <- lapply(tbl, function(u) {
  u_filtered <- subset(u, p_adj.loc < 0.3)
  u_filtered <- dplyr::arrange(u_filtered, p_adj.loc)
  return(u_filtered)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil, nrow, numeric(1))
p_de <- format(n_de / nrow(sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# calculate expression frequencies 
frq <- calcExprFreqs(sce, assay = "counts", th = 0)
# one sheet per cluster
assayNames(frq)
gids <- levels(sce$group_id)
frq10 <- vapply(as.list(assays(frq)), 
  function(u) apply(u[, gids] > 0.1, 1, any), 
  logical(nrow(sce)))
t(head(frq10))

```

```{r}
# second filtration step which can be done but not done for this analysis 
kids2 <- names(tbl_fil)[sapply(tbl_fil, nrow) != 0]
names(kids2) <- kids2

tbl_fil2 <- lapply(kids2, function(k)
  dplyr::filter(tbl_fil[[k]], 
    gene %in% names(which(frq10[, k]))))

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil2, nrow, numeric(1))
p_de <- format(n_de / nrow(sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

topfil_GEX <- do.call(rbind, tbl_fil)
top_GEX <- do.call(rbind, tbl)

# plot top table 
topfil_GEX %>% 
    ggplot(aes(x = logFC, y = -log10(p_val), color = cluster_id, label = gene)) +
  geom_point() +
  geom_text_repel() +
  theme_minimal() +
  labs(
    title = "Volcano Plot (Gene Expression)",
    x = "Log Fold Change",
    y = "-Log10 P-value",
    color = "Cluster")

```

## Visualize results 
```{r}
# between cluster concordance 
de_gs_by_k <- purrr::map(tbl_fil, "gene")
upset(fromList(de_gs_by_k))

# sample level viz: Pseudobulk heatmaps

# top-5 DS genes per cluster
pbHeatmap(sce, res, top_n = 5, )
```

```{r}
plotExpression(sce[, sce$cluster_id == "CB Granule Glut"],
  features = tbl_fil$`CB Granule Glut`$gene[seq_len(6)],
  x = "sample_id", colour_by = "group_id", ncol = 3) +
  guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

## Save top table 
```{r}
saveRDS(topfil_GEX, file = here::here("Data", "topfil_GEX.rds"))
saveRDS(top_GEX, file = here::here("Data", "top_GEX.rds"))
```

```{r}
sessionInfo()
```

